components/renku_data_services/connected_services/orm.py

```diff
@@ -0,0 +1,115 @@
+
+from datetime import datetime
+from typing import Any
+from urllib.parse import urljoin
+
+from sqlalchemy import JSON, DateTime, ForeignKey, MetaData, String, func
+from sqlalchemy.dialects.postgresql import JSONB
+from sqlalchemy.orm import DeclarativeBase, Mapped, MappedAsDataclass, mapped_column, relationship
+from sqlalchemy.schema import UniqueConstraint
+from ulid import ULID
+
+from renku_data_services import errors
+from renku_data_services.connected_services import models
+from renku_data_services.connected_services.apispec import ConnectionStatus, ProviderKind
+
+JSONVariant = JSON().with_variant(JSONB(), "postgresql")
+
+metadata_obj = MetaData(schema="connected_services")
+
+
+class BaseORM(MappedAsDataclass, DeclarativeBase):
+    """Base class for all ORM classes."""
+
+    metadata = metadata_obj
+
+
+class OAuth2ClientORM(BaseORM):
+    """An OAuth2 Client."""
+
+    __tablename__ = "oauth2_clients"
+    id: Mapped[str] = mapped_column("id", String(99), primary_key=True)
+    client_id: Mapped[str] = mapped_column("client_id", String(500))
+    display_name: Mapped[str] = mapped_column("display_name", String(99))
+    created_by_id: Mapped[str] = mapped_column("created_by_id", String())
+    kind: Mapped[ProviderKind]
+    scope: Mapped[str] = mapped_column("scope", String())
+    url: Mapped[str] = mapped_column("url", String())
+    client_secret: Mapped[str | None] = mapped_column("client_secret", String(500), default=None)
+    creation_date: Mapped[datetime] = mapped_column(
+        "creation_date", DateTime(timezone=True), default=None, server_default=func.now(), nullable=False
+    )
+    updated_at: Mapped[datetime] = mapped_column(
+        "updated_at",
+        DateTime(timezone=True),
+        default=None,
+        server_default=func.now(),
+        onupdate=func.now(),
+        nullable=False,
+    )
+
+    def dump(self, redacted: bool = True) -> models.OAuth2Client:
+        """Create an OAuth2 Client model from the OAuth2ClientORM."""
+        return models.OAuth2Client(
+            id=self.id,
+            kind=self.kind,
+            client_id=self.client_id if not redacted else "",
+            display_name=self.display_name,
+            scope=self.scope,
+            url=self.url,
+            created_by_id=self.created_by_id,
+            creation_date=self.creation_date,
+            updated_at=self.updated_at,
+        )
+
+    @property
+    def authorization_url(self) -> str:
+        """The authorization URL for the OAuth2 protocol."""
+        if not self.url:
+            raise errors.ValidationError(message=f"URL not defined for provider {self.id}.")
+        return urljoin(self.url, "oauth/authorize")
+
+    @property
+    def token_endpoint_url(self) -> str:
+        """The token endpoint URL for the OAuth2 protocol."""
+        if not self.url:
+            raise errors.ValidationError(message=f"URL not defined for provider {self.id}.")
+        return urljoin(self.url, "oauth/token")
```

Comment:
This feature does not care about OIDC, only about OAuth2, so I'm not sure why it would be useful to include code about OIDC here.

The PR for GitHub support updates these methods to adapt the URL based on the `ProviderKind`.